src/dmd/root/array.d

```diff
@@ -803,10 +803,262 @@ if (is(ReturnType!(typeof((T t) => callable(t))) == int))
     assert(returnValue == 7);
 }
 
-private template ReturnType(T)
+/// Returns: A static array constructed from `array`.
+pragma(inline, true) T[n] staticArray(T, size_t n)(auto ref T[n] array)
+{
+    return array;
+}
+
+///
+pure nothrow @safe @nogc unittest
+{
+    enum a = [0, 1].staticArray;
+    static assert(is(typeof(a) == int[2]));
+    static assert(a == [0, 1]);
+}
+
+/// Returns: `true` if the two given ranges are equal
+bool equal(Range1, Range2)(Range1 range1, Range2 range2)
+{
+    template isArray (T)
+    {
+        static if (is(T U : U[]))
+            enum isArray = true;
+
+        else
+            enum isArray = false;
+    }
+
+    static if (isArray!Range1 && isArray!Range2 && is(typeof(range1 == range2)))
+        return range1 == range2;
+
+    else
+    {
+        static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        {
+            if (range1.length != range2.length)
+                return false;
+        }
+
+        else
+        {
+            for (; !range1.empty; range1.popFront(), range2.popFront())
+            {
+                if (range2.empty)
+                    return false;
+
+                if (range1.front != range2.front)
+                    return false;
+            }
+        }
+
+        return true;
+    }
+}
+
+///
+pure nothrow @nogc @safe unittest
+{
+    enum a = [ 1, 2, 4, 3 ].staticArray;
+    static assert(!equal(a[], a[1..$]));
+    static assert(equal(a[], a[]));
+
+    // different types
+    enum b = [ 1.0, 2, 4, 3].staticArray;
+    static assert(!equal(a[], b[1..$]));
+    static assert(equal(a[], b[]));
+}
+
+/**
+ * Lazily filters the given range based on the given predicate.
+ *
+ * Returns: a range containing only elements for which the predicate returns
+ *  `true`
+ */
+auto filter(alias predicate, Range)(Range range)
+{
+    return Filter!(predicate, Range)(range);
+}
+
+///
+pure nothrow @safe @nogc unittest
+{
+    enum a = [1, 2, 3, 4].staticArray;
+    enum result = a[].filter!(e => e > 2);
+
+    enum expected = [3, 4].staticArray;
+    static assert(result.equal(expected[]));
+}
+
+private struct Filter(alias predicate, Range)
+{
+    private Range range;
+    private bool primed;
+
+    private void prime()
+    {
+        if (primed)
+            return;
+
+        while (!range.empty && !predicate(range.front))
+            range.popFront();
+
+        primed = true;
+    }
+
+    @property bool empty()
+    {
+        prime();
+        return range.empty;
+    }
+
+    @property auto front()
+    {
+        prime();
+        return range.front;
+    }
+
+    void popFront()
+    {
+        prime();
+
+        do
+        {
+            range.popFront();
+        } while (!range.empty && !predicate(range.front));
+    }
```

Comment:
I would add `opSlice` overloads, s.t. the ranges can be saved/copied (I see that you have already used this for built-in arrays)

Reply:
Done.